middle_end/flambda/lift_code.ml

```diff
@@ -17,102 +17,339 @@
+    (* [exclave (...)] *)
+
+(* A context composed from some number of [def]s. These are inside-first, so for
+   example (with abbreviated syntax),
+     [[Exclave; Immutable("x", 42); Mutable("y", "z"); Region]]
+   represents the fragment
+     [region (let mutable y = z in let x = 42 in exclave (...))]. *)
+type defs = def list
 
-let rebuild_let (defs : def list) (body : Flambda.t) =
+let rebuild_let (defs : defs) (body : Flambda.t) =
   let module W = Flambda.With_free_variables in
   List.fold_left (fun body def ->
     match def with
     | Immutable(var, def) ->
         W.create_let_reusing_defining_expr var def body
     | Mutable(var, initial_value, contents_kind) ->
-        Flambda.Let_mutable {var; initial_value; contents_kind; body})
+        Flambda.Let_mutable {var; initial_value; contents_kind; body}
+    | Region ->
+        Flambda.Region body
+    | Exclave ->
+        Flambda.Exclave body)
     body defs
 
-let rec extract_let_expr (acc:def list) (let_expr:Flambda.let_expr) :
-  def list * Flambda.t Flambda.With_free_variables.t =
+(* Whether [let x = region (... M) in N] can be rewritten as
+   [region (... let x = M in exclave N)]. Only true if [M] has no exclaves
+   (besides those in lambdas and nested regions) and no tail calls. (In the case
+   of tail calls, for [M] to have a tail call, the [N] in the above must be
+   simply [x] - in fact, it will be the special [lifted_let] variable created in
+   [lift_lets_expr] and we're relying on [split_defs] to substitute out the [x]
+   later.) *)
+let rec liftable_region_body0 (expr : Flambda.t) ~depth =
+  match expr with
+  | Exclave expr ->
+      depth > 0 (* Nested exclaves are fine *)
+      && liftable_region_body0 expr ~depth:(depth - 1)
+  | Apply { reg_close = Rc_close_at_apply; _ }
+  | Send { reg_close = Rc_close_at_apply; _ } ->
+      (* Tail calls must be kept tail calls *)
+      false
+  | Region expr ->
+      liftable_region_body0 expr ~depth:(depth + 1)
+  | Let { body; _ }
+  | Let_mutable { body; _ }
+  | Let_rec (_, body) ->
+      liftable_region_body0 body ~depth
+  | If_then_else (_, ifso, ifnot, _) ->
+      liftable_region_body0 ifso ~depth && liftable_region_body0 ifnot ~depth
+  | Switch (_, { consts; blocks; failaction; _ }) ->
+      liftable_cases0 consts ~depth
+      && liftable_cases0 blocks ~depth
+      &&
+      begin match failaction with
+      | None -> true
+      | Some expr -> liftable_region_body0 expr ~depth
+      end
+  | String_switch (_, cases, failaction, _) ->
+      liftable_cases0 cases ~depth
+      &&
+      begin match failaction with
+      | None -> true
+      | Some expr -> liftable_region_body0 expr ~depth
+      end
+  | Static_catch (_, _, body, handler, _)
+  | Try_with (body, _, handler, _) ->
+      liftable_region_body0 body ~depth && liftable_region_body0 handler ~depth
+  | Var _
+  | Apply _
+  | Send _
+  | Assign _
+  | Static_raise _
+  | While _
+  | For _
+  | Proved_unreachable ->
+    true
```

Comment:
This looks correct, but could it be done more compactly with something from Flambda_iterators? (The answer may be no here)

Reply:
I don't think so. What we're missing is a shallow version of `iter`, I think, so that we could perform the recursive call ourselves (changing the depth).

Reply:
Got it, thanks.